### DISCUSSION TITLE
Help with api?

### DISCUSSION THREAD

**Discussion by nexensys** [ScratchAddons/ScratchAddons#1528] [Help] [answered]
I want to create an addon, a script that will let users add blocks in a new way (not going to say any more here). Can someone help me understand the block injection process? I have a moderate knowledge of html css and javascript.

I am looking for something along the lines of:
```javascript
blocks.insert('block', 'data')
```
Or like what griffpatch's devtools does with the middle-click function...

Also, if possible, is it possible to check if another addon from the list is activated?
Thanks,
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ErrorGamer2000

**Comment by cobaltt7** [comment]
Just put all your code in between

```
export default async function({addon, global, msg , console}) {
    // your code here
}
________________________________
From: ErrorGamer2000 <notifications@github.com>
Sent: Monday, February 8, 2021 6:31:53 PM
To: ScratchAddons/ScratchAddons <ScratchAddons@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: [ScratchAddons/ScratchAddons] Help with api? (#1528)


I want to create an addon, a script that will let users add blocks in a new way (not going to say any more here). Can someone help me understand the block injection process and what the whole "export default" bit of all of the userscripts are? I have a moderate knowledge of html css and javascript.

Thanks,
      ErrorGamer2000

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/ScratchAddons/ScratchAddons/discussions/1528>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT5DEVDQMPKOCEBTY7E3F3S6B67TANCNFSM4XJ75KGA>.

**Reply by nexensys** [reply]
I was pretty sure about that, and just wanted to check. Thanks! Do you have any idea about injecting blocks?

**Reply by cobaltt7** [reply]
also, the `{addon, global, msg, console}` part can be changed depending on your needs.
[if you include `addon` you can use](https://github.com/ScratchAddons/ScratchAddons/wiki/addon.*-APIs-reference)
[if you include `global` you can do](https://github.com/ScratchAddons/ScratchAddons/wiki/About-persistent-scripts#technical-aspects-of-persistent-scripts:~:text=global%3A%20this%20is%20a%20shared%20object,%7D)
[if you include `msg` (or `safeMsg` if you are injecting it directly into the HTML) you can do](https://github.com/ScratchAddons/ScratchAddons/wiki/Localizing-addons)
always include `console`, that way your `console.log()`s will say what addon it's from

**Comment by cobaltt7** [comment]
Injecting blocks is a rejected addon. Non extension users wouldn't be able to use them.

But no,I have no idea how to do that.
________________________________

**Reply by nexensys** [reply]
Ok, thanks anyway!

**Reply by pufferfish101007** [reply]
Injecting blocks is not rejected... creating new blocks that _didn't already exist_ and injecting is rejected.

**Reply by cobaltt7** [reply]
That's what I meant, sorry I didn't clarify.

**Comment by pufferfish101007** [comment, marked as answer]
To inject blocks, you'll need to use the scratch-blocks library (include it in your addon.json libraries) which can then be used as Blockly in your code. What you can do is construct a block's xml and then call the paste() function on the workspace you're injecting on. I think you can usually get it with Blockly.getMainWorkspace(), but I may be wrong. Take a look at https://github.com/ScratchAddons/ScratchAddons/blob/master/addons/block-switching/userscript.js for an example of using the workspace.paste method.

**Reply by nexensys** [reply]
Thanks!

**Comment by GrahamSH-LLK** [comment]
We're adding a new API (addon.addBlock) that will be merged with #2161
